Add accessibility to CompassView and NearbyTableViewCell

hhanke opened the pull request:
Treat the whole NearbyTableViewCell as one accessibility element and add distance and direction descriptions to its label.

bgerstle commented on Wikipedia/Code/WMFCompassView.m:
```diff
@@ -223,4 +223,21 @@ - (void)drawCompassLinesInContext:(CGContextRef)ctx
+}
+
+- (NSString *)accessibilityLabel {
+    double positive_angle = self.angleRadians >= 0 ? self.angleRadians : self.angleRadians + 2*3.14;
```
please use math constants, e.g. `2_M_PI` for 2 * pi.

bgerstle commented on Wikipedia/Code/WMFCompassView.m:
```diff
@@ -223,4 +223,21 @@ - (void)drawCompassLinesInContext:(CGContextRef)ctx
+        clock_direction = 12;
+    }
+    NSString* label = MWLocalizedString(@"compass-direction", nil);
+    label = [label stringByReplacingOccurrencesOfString:@"$1" withString:[NSString localizedStringWithFormat:@"%@", @(clock_direction)]];
```
i think the `localizedStringWithFormat` is unnecessary here, since all the values are "**N** o'clock" and don't vary based on units of measurement or 12/24-hour time.

bgerstle replied:
also, do you have any sense for whether the "o'clock" convention is used across languages?  for example, how would you convey directionality in Hebrew, German, or Spanish?

coreyfloyd replied:
yeah this may be a time where we want to use the base units of degrees which is universal. Though not as nice in english clock positions, it should be understood by everyone.

bgerstle replied:
i did a bit of research by turning on VoiceOver and using the built-in Compass app. It uses degrees for all languages I tested (English, Spanish, & Chinese), plus the initials the cardinal direction, e.g. 145 degrees, SE (it says "S-E" not "Southeast").  So I'd suggest:

- Converting radians to degrees (using `RADIANS_TO_DEGREES`)
- Optionally, adding cardinal direction (N, NE, E, SE, S, etc.)
- Localizing the label using the format: "$1 degrees" and, optionally, combining that w/ another localized string for the direction initials (i.e. `MWLocalizedString(@"west-initial", nil)` would be "W" in English, and "O" in Spanish)

bgerstle replied:
Side note, something interesting about our "compass" view is that it doesn't even tell you where North is. In fact, the "arrow" of our compass points in the direction of the object, not North (as is the convention, AFAIK).

bgerstle replied:
@Hhanke @coreyfloyd for now, let's just use degrees and we can follow-on the cardinal direction later if needed

hhanke replied:
Regarding the clock direction system, this is widely used among blind people also outside of English speaking countries. I can definitely confirm that for my country, the Czech Republic. 

It is being used for description of direction in orientation (e.g. when you describe to a blind person where objects are around him), in navigation (e.g. the Blindsquare iOS app) or e.g. to describe the position of items on someones plate while eating.

But Brian is right, we need to check how this is handled in more remote countries like China. I get a feeling that the clockface is very widely known. I will try to check it and I will write another comment here.

I think at least in countries where it is routinely used among blind people, it is much superior to giving degrees in a situation like this. Degrees carry too much detail and are complicated to imagine. I do not know how about you, but I cannot intuitively (without some thinking) realize what a direction like like 210 degrees, 72 degrees or 315 degrees actually mean around me.

The Compass app is not a good example. In that case, the aim is to give an exact direction and that is the principal purpose of the app. In the Wikipedia app, if I understand correctly, the main purpose is giving the users easy access to articles containing information about their surroundings. The information about distance and direction in this later case serve another purpose and should be easily understandable.

So maybe we will have to differentiate based on the current language settings and maybe default to degrees where a better system is not available.

For the same reason, I also think that the cardinal direction is not really necessary, although it could be a nice addition and might also be relevant in some cases.

Regarding the first comment, isn't localizedStringWithFormat necessary for languages which use other characters for numerals, e.g. Japanese?

Thanks for the other comments as well!

bgerstle replied:
> Regarding the first comment, isn't localizedStringWithFormat necessary for languages which use other characters for numerals, e.g. Japanese?

Oh, didn't consider that... we might need to do that in some other places as well.  For example, using the Swift REPL to print a localized string arabic numeral(?):

```
 8> NSString(format: "%@", locale:NSLocale(localeIdentifier:"ar"), NSNumber(int: 1))
$R6: NSString = "١"
  9> NSString(format: "%d", locale:NSLocale(localeIdentifier:"ar"), 1)
$R7: NSString = "١"
```

bgerstle commented on Wikipedia/Code/WMFNearbyArticleTableViewCell.m:
```diff
@@ -197,4 +197,20 @@ - (void)setImage:(MWKImage*)image {
+    } else {
+        titleAndDescription = self.titleText;
+    }
+    return [NSString stringWithFormat: @"%@, %@ %@", titleAndDescription, self.distanceLabel.text, self.compassView.accessibilityLabel];
```
maybe a minor point, but shouldn't the accessibility label for the distance say the full unit name, i.e. meters/feet instead of m/ft?

bgerstle commented on Wikipedia/Code/WMFCompassView.m:
```diff
@@ -223,4 +223,21 @@ - (void)drawCompassLinesInContext:(CGContextRef)ctx
+
+- (NSString *)accessibilityLabel {
+    double positive_angle = self.angleRadians >= 0 ? self.angleRadians : self.angleRadians + 2*3.14;
+    NSInteger clock_direction = lroundf(positive_angle / (3.14 / 6));
```
would you mind writing unit tests for these calculations? e.g. that some example angles return the expected "clock" orientation

coreyfloyd replied:
see my not above about degrees before investing time in a unit test.

bgerstle commented:
@hhanke i've sent an email to some mailing lists asking for advice on the issue. in the meantime can we go with whatever is simplest to implement?  i agree with you that degrees and cardinal directions aren't ideal for this use case, so maybe we can go w/ clock faces while we look into the issue further?

hhanke commented:
Thank you Brian! I have also sent some questions to people whom I think might know this. Should we then wait a few days to see what will be the feedback?

bgerstle commented:
@hhanke we could, but IIRC there are other worthwhile changes in here that don't need to be delayed.  even if the compass itself isn't accessible yet (or says degrees, or something) it will be an improvement.  either way, I'd prefer to not leave this open while the issue is pending (we can have a ticket on the backlog)

hhanke commented:
@bgerstle OK, I will prepare it. We know that for people from some countries, the clockface is right. If it turns out that we need to do more from speakers from other countries, we can always improve it.

bgerstle commented:
> we can always improve it

that's the idea 😊

hhanke commented:
I have made the conversion from radians to degrees more robust, moved it to WMFMath and provided a test for it. I have also used the M_PI constant and fixed the naming of variables. It would be nicer to use "20 meters" instead of "20 m", but presently, it has the same effect as VoiceOver expands measurements shortcuts during TTS anyway. So it does in fact read "20 meters" to the user. Therefore, I didn't change this part and maybe it falls into the category of what can be improved later on :-)

bgerstle commented:
👍🏻

coreyfloyd commented:
looking good - thanks @hhanke!
